Limit loop unrolling

aorenste opened the pull request:
Tacotron2 causes massive loop unrolling resulting in very large graphs (26k nodes) which was causing inductor (and tracing itself) to choke.

The unrolling size is controlled by the environment variable TORCHDYNAMO_MAX_LOOP_UNROLL_NODES which defaults to the arbitrary value 5000.

This updates the tacotron2 timings as follows:
eager timing: 3m:23s -> 35s
aot_eager timing: 4m:12s -> 39s
inductor timing: 22m:24s ->1m

For reference the big loop in tacotron2 was this one (model.py[405]):
```
        while len(mel_outputs) < decoder_inputs.size(0) - 1:
            decoder_input = decoder_inputs[len(mel_outputs)]
            mel_output, gate_output, attention_weights = self.decode(decoder_input)
            mel_outputs += [mel_output.squeeze(1)]
            gate_outputs += [gate_output.squeeze(1)]
            alignments += [attention_weights]
```
which gets unrolled and inlined adding about 36 nodes to the graph per iteration.

Fixes #98467 
Relates to #102839 which hopefully will result in a better fix.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120023



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

aorenste committed:
Limit loop unrolling

[ghstack-poisoned]

aorenste committed:
Update on "Limit loop unrolling"

Tacotron2 causes massive loop unrolling resulting in very large graphs (26k nodes) which was causing inductor (and tracing itself) to choke.

This updates the tacotron2 timings as follows:
eager timing: 3m:23s -> 35s
aot_eager timing: 4m:12s -> 39s
inductor timing: 22m:24s ->1m






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

aorenste committed:
Update on "Limit loop unrolling"

Tacotron2 causes massive loop unrolling resulting in very large graphs (26k nodes) which was causing inductor (and tracing itself) to choke.

This updates the tacotron2 timings as follows:
eager timing: 3m:23s -> 35s
aot_eager timing: 4m:12s -> 39s
inductor timing: 22m:24s ->1m

For reference the big loop in tacotron2 was this one (model.py[405]):
```
        while len(mel_outputs) < decoder_inputs.size(0) - 1:
            decoder_input = decoder_inputs[len(mel_outputs)]
            mel_output, gate_output, attention_weights = self.decode(decoder_input)
            mel_outputs += [mel_output.squeeze(1)]
            gate_outputs += [gate_output.squeeze(1)]
            alignments += [attention_weights]
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

aorenste commented:
I'm thinking this should probably include a user-facing release note since it could potentially change visible behavior (causing a loop that was previously not graph breaking to become graph breaking)...

oulgen commented on torch/_dynamo/config.py:
```diff
@@ -401,6 +401,11 @@ def default_debug_dir_root():
+# Approximate maximum number of nodes to unroll loops into. A value of 0 will
+# unroll fully but could result in very large graphs that take an inordinate
+# amount of time to process.
+max_loop_unroll_nodes = int(os.environ.get("TORCHDYNAMO_MAX_LOOP_UNROLL_NODES", 5000))
```
On the compile time improvements doc, I wrote down
```
[With Horace] Partial graphs
Instead of emitting one fully inlined graph, have dynamo emit particle graphs
For example, if you have a large for loop, instead of unrolling it, emit a partial graph for its contents, and re-use it. Balance this between compile time versus performance.
```

This sounds like a perfect example for when something like this would really benefitial.

@Chillee @anijain2305

yanboliang replied:
The same idea at https://github.com/pytorch/pytorch/issues/102839#issuecomment-1638873591

oulgen commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -581,6 +581,34 @@ def handle_graph_break(
+        # like watching the rate of growth to trim loops earlier (so we don't
+        # have to wait for `max` nodes before skipping).
+        if config.max_loop_unroll_nodes > 0 and count > config.max_loop_unroll_nodes:
+            raise exc.SkipFrame("unrolled loop getting too big")
```
Can you explain how this fixes the problem? When we throw this exception we just graph break and fall back to eager mode?

aorenste replied:
That's right. This doesn't make the underlying problem go away - it just makes the pathological case not misbehave.

aorenste replied:
I'm thinking maybe I should change this subtly so instead of checking for total nodes it checks for nodes added by this particular backedge (so 1 loop expanding into 6000 nodes is not expanded but 4 independent loops expanding into 2000 nodes each is fine). The point is to be a backstop - not a hard limit.

aorenste replied:
Also maybe this should treat full_graph mode as '0' as well...

yanboliang replied:
> Can you explain how this fixes the problem? When we throw this exception we just graph break and fall back to eager mode?

@oulgen Actually this is not graph break, this is skip frame, which means the whole frame doesn't get compiled.

yanboliang commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -581,6 +581,34 @@ def handle_graph_break(
+
+        # Don't skip if we haven't seen this particular backedge at least a few
+        # times.
+        if self.n_seen < 3:
```
Probably need a slightly large threshold, I suspect internal Ads model may exceed 3.

aorenste committed:
Update on "Limit loop unrolling"

Tacotron2 causes massive loop unrolling resulting in very large graphs (26k nodes) which was causing inductor (and tracing itself) to choke.

The unrolling size is controlled by the environment variable TORCHDYNAMO_MAX_LOOP_UNROLL_NODES which defaults to the arbitrary value 5000.

This updates the tacotron2 timings as follows:
eager timing: 3m:23s -> 35s
aot_eager timing: 4m:12s -> 39s
inductor timing: 22m:24s ->1m

For reference the big loop in tacotron2 was this one (model.py[405]):
```
        while len(mel_outputs) < decoder_inputs.size(0) - 1:
            decoder_input = decoder_inputs[len(mel_outputs)]
            mel_output, gate_output, attention_weights = self.decode(decoder_input)
            mel_outputs += [mel_output.squeeze(1)]
            gate_outputs += [gate_output.squeeze(1)]
            alignments += [attention_weights]
```
which gets unrolled and inlined adding about 36 nodes to the graph per iteration.

Fixes #98467 
Relates to #102839 which hopefully will result in a better fix.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

aorenste committed:
Update on "Limit loop unrolling"

Tacotron2 causes massive loop unrolling resulting in very large graphs (26k nodes) which was causing inductor (and tracing itself) to choke.

The unrolling size is controlled by the environment variable TORCHDYNAMO_MAX_LOOP_UNROLL_NODES which defaults to the arbitrary value 5000.

This updates the tacotron2 timings as follows:
eager timing: 3m:23s -> 35s
aot_eager timing: 4m:12s -> 39s
inductor timing: 22m:24s ->1m

For reference the big loop in tacotron2 was this one (model.py[405]):
```
        while len(mel_outputs) < decoder_inputs.size(0) - 1:
            decoder_input = decoder_inputs[len(mel_outputs)]
            mel_output, gate_output, attention_weights = self.decode(decoder_input)
            mel_outputs += [mel_output.squeeze(1)]
            gate_outputs += [gate_output.squeeze(1)]
            alignments += [attention_weights]
```
which gets unrolled and inlined adding about 36 nodes to the graph per iteration.

Fixes #98467 
Relates to #102839 which hopefully will result in a better fix.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

aorenste committed:
Update on "Limit loop unrolling"

Tacotron2 causes massive loop unrolling resulting in very large graphs (26k nodes) which was causing inductor (and tracing itself) to choke.

The unrolling size is controlled by the environment variable TORCHDYNAMO_MAX_LOOP_UNROLL_NODES which defaults to the arbitrary value 5000.

This updates the tacotron2 timings as follows:
eager timing: 3m:23s -> 35s
aot_eager timing: 4m:12s -> 39s
inductor timing: 22m:24s ->1m

For reference the big loop in tacotron2 was this one (model.py[405]):
```
        while len(mel_outputs) < decoder_inputs.size(0) - 1:
            decoder_input = decoder_inputs[len(mel_outputs)]
            mel_output, gate_output, attention_weights = self.decode(decoder_input)
            mel_outputs += [mel_output.squeeze(1)]
            gate_outputs += [gate_output.squeeze(1)]
            alignments += [attention_weights]
```
which gets unrolled and inlined adding about 36 nodes to the graph per iteration.

Fixes #98467 
Relates to #102839 which hopefully will result in a better fix.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

aorenste commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot revert -m "breaks llms export"

anijain2305 commented:
@pytorchbot revert -m "breaks llms export" -c nosignal

aorenste commented:
This was reverted because the arbitrary node limit was deemed unacceptable.  I can pretty easily change it to a loop count limit (or some combination) and add a fullgraph check - but I fear that the same concerns will still apply (it's an arbitrary limit).

Thoughts?
